keyvault/certrefresher.go

```diff
@@ -0,0 +1,130 @@
+package keyvault
+
+import (
+	"context"
+	//nolint:gosec // sha1 only used to display cert thumbprint in logs for cross-verification with keyvault.
+	"crypto/sha1"
+	"crypto/tls"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/avast/retry-go/v3"
+	"github.com/pkg/errors"
+)
+
+type EventualExpirationErr struct {
+	time.Time
+}
+
+func (e *EventualExpirationErr) Error() string {
+	return fmt.Sprintf("could not refresh before expiration on %s", e.Time.String())
+}
+
+type tlsCertFetcher interface {
+	GetLatestTLSCertificate(ctx context.Context, certName string) (tls.Certificate, error)
+}
+
+type logger interface {
```

Comment:
I switched from `zap` to existing logging mechanism to keep any possible side effects from this PR to a minimum. I enjoyed adding logs with `zap` so when given a chance I plan to revisit this package (or CNS in general) to contribute to its adoption.